Reduce log noise on SELinux mount mismatch

jsafrane opened the pull request:
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
The Desired State of World can require a different SELinux mount context than is in the Actual State of World and it's perfectly OK. For example when user changes SELinux context of Pods or when the context is reconstructed after kubelet restart.

Don't spam log and don't report errors to the user as event - reconciler will do the right thing and unmount the old volume (with wrong context) and mount a new one in the next reconciliation. It's not an error, it's expected workflow.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related to https://github.com/kubernetes/enhancements/pull/3548

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/3548
```

jsafrane committed:
Reduce log noise on SELinux mount mismatch

The Desired State of World can require a different SELinux mount context than
is in the Actual State of World and it's perfectly OK. For example when
user changes SELinux context of Pods or when the context is reconstructed
after kubelet restart.

Don't spam log and don't report errors to the user as event - reconciler
will do the right thing and unmount the old volume (with wrong context) and
mount a new one in the next reconciliation. It's not an error, it's
expected workflow.

msau42 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -220,10 +220,10 @@ func (rc *reconciler) mountOrAttachVolumes() {
-			klog.ErrorS(err, volumeToMount.GenerateErrorDetailed("mount precondition failed, please report this error in https://github.com/kubernetes/enhancements/issues/1710, together with full Pod yaml file", err).Error(), "pod", klog.KObj(volumeToMount.Pod))
-			// TODO: report error better, this may be too noisy
-			rc.desiredStateOfWorld.AddErrorToPod(volumeToMount.PodName, err.Error())
+			// The volume is mounted, but with an unexpected SELinux context.
```
Do you think a metric would be useful?

jsafrane replied:
No, a metric would  grow like crazy, this is called every 100ms. In addition, it is a part of the "good" workflow - similar to `IsVolumeNotAttachedError` below. Kubelet must unmount the old volume (with wrong SELinux label) first.

jsafrane replied:
Now that I think about it, calling `AddErrorToPod` still could be useful, in case unmount of the other volume takes too long.

jsafrane replied:
Added a separate commit with `AddErrorToPod`.

Regarding the metric - there already is a metric when a second pod with SELinux label that does not match the already present one is rejected and that's an error. To get a metric when a volume is being unmounted because of unexpected SELinux label, I would need to add something into `unmountDetachDevices`...

jsafrane commented:
/retest

flakes in `vendor/k8s.io/apiserver/pkg/storage/`

jsafrane committed:
Report error on a pod startup on SELinux mismatch

When a volume is already mounted with an unexpected SELinux label,
kubelet must unmount it first and then mount it back with the expected one.
Report an error to user, just in case the unmount takes too long.

In therory, this error should not happen too often, because two Pods with
different SELinux label will not enter Desired State of World, see
dsw.AddPodToVolume. It can happen when DSW and ASW SELinux labels only when
a volume has been deleted from DSW (= Pod was deleted) or a volume was
reconstructed after kubelet restart. In both cases, volume manager should
unmount the volume quickly.
